src/geofabrics/bathymetry_estimation.py

```diff
@@ -0,0 +1,1576 @@
+        """ Create a channel object from a REC file.
+
+        Parameters
+        ----------
+
+        rec_network
+            Contains association information between upstream and
+            downstream reaches.
+        reach_id
+            The name of the reach ID in the REC channel.
+        area_threshold
+            The area threshold in metres squared below which to ignore a reach.
+        max_iterations
+            The maximum number of iterations along a single strand to trace
+            upstream.
+        iteration
+            The number of iterations traveled upstream.
+        """
+        reaches, iteration = cls._get_up_stream_reaches(rec_network=rec_network,
+                                                        reach_id=reach_id,
+                                                        reaches=None,
+                                                        max_iterations=max_iterations,
+                                                        iteration=0)
+        reaches = reaches[reaches['CUM_AREA'] > area_threshold]
+        sampling_direction = -1
+        channel = cls(channel=reaches,
+                      resolution=resolution,
+                      sampling_direction=sampling_direction)
+        return channel
+
+    @classmethod
+    def _get_up_stream_reaches(cls,
+                               rec_network: geopandas.GeoDataFrame,
+                               reach_id: int,
+                               reaches: geopandas.GeoDataFrame,
+                               max_iterations: int,
+                               iteration: int):
+        """ A recurive function to trace all up reaches from the reach_id.
+        The default values for reaches and iteration are set for the
+        initial call to the recursive function.
+
+        Parameters
+        ----------
+
+        rec_network
+            Contains association information between upstream and
+            downstream reaches.
+        reach_id
+            The `nzsegment` id of the reach to trace upstream from.
+        reaches
+            The already traced downstream reaches to append to.
+        max_iterations
+            The maximum number of iterations along a single strand to trace
+            upstream.
+        iteration
+            The number of iterations traveled upstream.
+        """
+        if reaches is None:
+            reaches = rec_network[rec_network['nzsegment'] == reach_id]
+        if iteration > max_iterations:
+            print(f"Reached recursion limit at: {iteration}")
+            return reaches, iteration
+        iteration += 1
+        up_stream_reaches = rec_network[rec_network['NextDownID']
+                                        == reach_id]
+        reaches = reaches.append(up_stream_reaches)
+        for index, up_stream_reach in up_stream_reaches.iterrows():
+            if not up_stream_reach['Headwater']:
+                reaches, iteration = cls._get_up_stream_reaches(
```

Comment:
I tend to avoid recursion because it can be memory intensive and sometimes hard to read. This isn't too hard to read but iteration may be clearer. I would consider using iteration instead, but if there are no problems with memory then you don't necessarily need to